### ISSUE TITLE
ak.concatenate tweaks

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/awkward#545]
First, `ak.concatenate`'s call of `to_layout` should have `allow_other=True`:

```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), 999], axis=1)
```

should return `[[1, 2, 3, 999], [999], [4, 5, 999]]`, not raise TypeError.

Second, the UnionArray that it creates should be simplified (`out.simplify(mergebool=False)`):

```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), ak.Array([[123], [223], [323]])], axis=1)
<Array [[1, 2, 3, 123], [223], [4, 5, 323]] type='3 * var * union[int64, int64]'>
```

The type should be `3 * var * int64`.

Finally, I don't know why it's not broadcasting here:

```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), ak.Array([123, 223, 323])], axis=1)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/irishep/awkward-1.0/awkward1/operations/structure.py", line 1020, in concatenate
    out = awkward1._util.broadcast_and_apply(contents, getfunction,
  File "/home/jpivarski/irishep/awkward-1.0/awkward1/_util.py", line 996, in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
  File "/home/jpivarski/irishep/awkward-1.0/awkward1/_util.py", line 759, in apply
    outcontent = apply(nextinputs, depth + 1)
  File "/home/jpivarski/irishep/awkward-1.0/awkward1/_util.py", line 577, in apply
    function = getfunction(inputs, depth)
  File "/home/jpivarski/irishep/awkward-1.0/awkward1/operations/structure.py", line 1015, in getfunction
    out = inputs[0].mergemany_as_union(inputs[1:], 1)
ValueError: axis out of range for flatten

(https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/libawkward/array/NumpyArray.cpp#L1431)
```

I would expect the result `[[1, 2, 3, 123], [223], [4, 5, 323]]`.

**Comment by ianna** [comment]
Ha it didn't take long :-)

**Comment by jpivarski** [comment]
Really, these are things I should have checked when reviewing your code, but I didn't think of them. They came up in use-cases (questions on Slack).

**Comment by ianna** [comment]
@jpivarski - `concatenate` currently expects a `Content` subtype as an argument (in any axis). Shall it be extended to accept other types?

for example, this fails with a `ValueError: content argument must be a Content subtype`
```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), 999], axis=0)
```

**Comment by jpivarski** [comment]
Just before entering the C++, `999` should be promoted to a NumpyArray:

```python
>>> length = 10
>>> awkward1.layout.NumpyArray(numpy.repeat(999, length))
<NumpyArray format="l" shape="10" data="999 999 999 999 999 999 999 999 999 999" at="0x5598403ae950"/>
```

At that point in the process, you'll know what `length` is supposed to be.

**Comment by ianna** [comment]
should this operation be commutative?
```python
ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), 999], axis=1) == ak.concatenate([999, ak.Array([[1, 2, 3], [], [4, 5]])], axis=1)
```

**Comment by jpivarski** [comment]
No, because in the first, the `999` goes at the end of each list, and in the second, the `999` goes at the beginning.

But this is already the case:

```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), ak.Array([[999], [999], [999]])], axis=1)
<Array [[1, 2, 3, 999], [999], [4, 5, 999]] type='3 * var * union[int64, int64]'>

>>> ak.concatenate([ak.Array([[999], [999], [999]]), ak.Array([[1, 2, 3], [], [4, 5]])], axis=1)
<Array [[999, 1, 2, 3], [999], [999, 4, 5]] type='3 * var * union[int64, int64]'>
```

**Comment by ianna** [comment]
Should it be allowed to concatenate in different axis? The second array has only `axis=0`
```python
>>> ak.concatenate([ak.Array([[1, 2, 3], [], [4, 5]]), ak.Array([123, 223, 323])], axis=1)
```

**Comment by jpivarski** [comment]
`broadcast_and_apply` adds a dimension as it descends so that when you get to your `getfunction`, they have the same dimension.

**Comment by jpivarski** [comment]
Fixed by #548.
